tests/docs/test_run_generate_dataset_example.py

```diff
@@ -0,0 +1,91 @@
+import textwrap
+import unittest
+
+import pytest
+
+from . import test_complete_examples
+
+
+def test_run_generate_dataset_example(tmp_path):
+    example = test_complete_examples.DatasetDefinitionExample(
+        path="test",
+        fence_number=1,
+        source=textwrap.dedent(
+            """\
+            from ehrql import create_dataset
+            from ehrql.tables.beta.tpp import patients
+
+            dataset = create_dataset()
+            dataset.define_population(patients.exists_for_patient())
+            """
+        ),
+    )
+    test_complete_examples.test_ehrql_generate_dataset_example(tmp_path, example)
+
+
+def test_run_generate_dataset_example_failing(tmp_path):
+    example = test_complete_examples.DatasetDefinitionExample(
+        path="test",
+        fence_number=1,
+        source=textwrap.dedent(
+            """\
+            from ehrql import create_dataset
+
+            dataset = create_dataset()
+            dataset.define_population(not_a_function())
+            """
+        ),
+    )
+    with pytest.raises(test_complete_examples.DatasetDefinitionTestError) as exc_info:
+        test_complete_examples.test_ehrql_generate_dataset_example(tmp_path, example)
+    assert type(exc_info.value) is test_complete_examples.DatasetDefinitionTestError
+
+
+def test_run_generate_dataset_example_failing_codelist_from_csv_call(tmp_path):
+    example = test_complete_examples.DatasetDefinitionExample(
+        path="test",
+        fence_number=1,
+        source=textwrap.dedent(
+            """\
+            from ehrql import codelist_from_csv, create_dataset
+            from ehrql.tables.beta.tpp import patients
+
+            codes = codelist_from_csv()
+
+            dataset = create_dataset()
+            dataset.define_population(patients.exists_for_patient())
+            """
+        ),
+    )
+
+    with pytest.raises(
+        test_complete_examples.DatasetDefinitionTestError,
+        match=r"generate_dataset failed for example",
+    ) as exc_info:
+        test_complete_examples.test_ehrql_generate_dataset_example(tmp_path, example)
+    assert type(exc_info.value) is test_complete_examples.DatasetDefinitionTestError
+
+
+def test_run_generate_dataset_example_gives_unreadable_csv(tmp_path):
+    example = test_complete_examples.DatasetDefinitionExample(
+        path="test",
+        fence_number=1,
+        source=textwrap.dedent(
+            """\
+            from ehrql import create_dataset
+            from ehrql.tables.beta.tpp import patients
+
+            dataset = create_dataset()
+            dataset.define_population(patients.exists_for_patient())
+            """
+        ),
+    )
+    with unittest.mock.patch("ehrql.main.generate_dataset", return_value=None):
+        with pytest.raises(
+            test_complete_examples.DatasetDefinitionTestError,
+            match=r"Check of output dataset CSV failed for example",
```

Comment:
question: I actually wanted to use [exception notes as added to Python 3.11](https://docs.python.org/3/tutorial/errors.html) in place of a custom exception to add these messages as notes to the original exception.

pytest [apparently supports these](https://github.com/pytest-dev/pytest/issues/11223) including using `match` to check against the notes. But I was finding that pytest was only seeing the first line of a `FileNotFoundError` exception which I was trying to raise.

```
AssertionError: Regex pattern did not match.                                                                                                                                                                 
 Regex: 'Check of output dataset CSV failed for example'                                                                                                                                                     
 Input: "[Errno 2] No such file or directory: /tmp/… 
```

I don't know if I'm doing something wrong in how I was trying to raise the exception. In [this code](https://github.com/opensafely-core/ehrql/pull/1648/commits/e4b718eb4a683b31b2007329e86c3c4e18a4beb3), I was doing:

```python
try:
    validate_dataset_output(tmp_dataset_path)
except Exception as e:
    e.add_note(f"Check of output dataset CSV failed for example: \n{formatted_example}")
    raise
```

If anyone think exception notes are a better approach and knows what I'm doing wrong, I'd love to know!